include/zephyr/linker/common-rom/common-rom-api.ld

```diff
@@ -0,0 +1,7 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <zephyr/linker/iterable_sections.h>
+
+#if defined(CONFIG_ADC)
+	ITERABLE_SECTION_ROM(adc_driver_api, Z_LINK_ITERABLE_SUBALIGN)
+#endif
```

Comment:
note: this means oot drivers will also need to create new linker files, and they won't be able to add stuff to `common-rom.ld`. Would it make sense to move this to each driver class folder? So in-tree/oot experience would be the same. Also, a cmake helper tied to the linker generator calls would be nice to have.

Reply:
> note: this means oot drivers will also need to create new linker files

Yes

> and they won't be able to add stuff to `common-rom.ld`. 

Does this matter? As long as it's located in ROM, it should be fine, no?

> Would it make sense to move this to each driver class folder? So in-tree/oot experience would be the same. 

This can be done if desired

> Also, a cmake helper tied to the linker generator calls would be nice to have.

A bit out of scope here, also note that the order of linker sections is important (which wouldn't be an issue here) and moving it to a cmake helper would make this harder to follow.

Reply:
> > note: this means oot drivers will also need to create new linker files
> 
> Yes, if they add an API other than those in-tree
> 
> > and they won't be able to add stuff to `common-rom.ld`.
> 
> Does this matter? As long as it's located in ROM, it should be fine, no?

IIRC there was a reason for this file, if there's no longer a reason, better keep stuff close to the code it belongs to (drivers/adc).

> 
> > Would it make sense to move this to each driver class folder? So in-tree/oot experience would be the same.
> 
> This can be done if desired

We should offer a consistent experience, so yes. `example-application` now contains a custom class, it can be used as a test bed.

> 
> > Also, a cmake helper tied to the linker generator calls would be nice to have.
> 
> A bit out of scope here, also note that the order of linker sections is important (which wouldn't be an issue here) and moving it to a cmake helper would make this harder to follow.

What I meant here is that this will need to work with the linker generator stuff, ie, no ld files.

Reply:
Moved it to the `drivers/adc` directory.

Reply:
> What I meant here is that this will need to work with the linker generator stuff, ie, no ld files.

Sure, but it would still need to end up in `linker.cmd` somewhere.

Reply:
I think CI shows why the `common-rom.ld` is used. If the device has no flash, it can't use the `SECTIONS` linker location.

Reply:
@cfriedt do you have some context on that?

Reply:
I've created #71911 to allow adding linker sections to ROM (similar to how `common-rom.ld` puts a lot of structs in ROM)

We might even use it to move some of clutter out of all the `common-rom-***.ld` scripts.

Reply:
Moved to the `gen_iter_sections.py` script.